### PR TITLE
Add finetuning strategies for DeepSpeed

### LINKED ISSUE #1249: Flash DeepSpeedPlugin error
## 🐛 Bug
Hi, I'm completely new Flash, I'm having this error when trying to use DeepSpeedPlugin plugin on the Flash trainer.
The code is 

### To Reproduce

```python

import torch

import flash
from flash.audio import SpeechRecognition, SpeechRecognitionData
from flash.core.data.utils import download_data

# 1. Create the DataModule
download_data("https://pl-flash-data.s3.amazonaws.com/timit_data.zip", "./data")

datamodule = SpeechRecognitionData.from_json(
    "file",
    "text",
    train_file="data/timit/train.json",
    test_file="data/timit/test.json",
    batch_size=4,
)

# 2. Build the task
model = SpeechRecognition(backbone="facebook/wav2vec2-base-960h")

# 3. Create the trainer and finetune the model
trainer = flash.Trainer(max_epochs=1, gpus=torch.cuda.device_count(),plugins='deepspeed')
trainer.finetune(model, datamodule=datamodule, strategy="no_freeze")

# 4. Predict on audio files!
datamodule = SpeechRecognitionData.from_files(predict_files=["data/timit/example.wav"], batch_size=4)
predictions = trainer.predict(model, datamodule=datamodule)
print(predictions)

# 5. Save the model!
trainer.save_checkpoint("speech_recognition_model.pt")
```

I'm getting this
Screenshot 2022-03-25 at 16 45 29

I thing there is an issue with named_params somewhere

### PR DESCRIPTION
**Pull request by ar90n**
## What does this PR do?

This PR provides some workarounds to use DeepSpeed in finetuning. In fact, DeepSpeed cannot work with pytorch-lightning completely because its parameter loading and storing don't work. So this PR added some fine-tuning strategies whose parameter loading and storing are omitted. 

Fixes #1249

## Before submitting
- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [x] Did you verify **new and existing tests pass** locally with your changes?
- [x] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ar90n** [comment]
I don't know how to fix `Lightning-AI.lightning-flash (Examples)` jobs. Could you give me some help?

File: docs/source/general/finetuning.rst
Comment (krshrimali) [inline docs/source/general/finetuning.rst:253]: Thanks for adding this section here! 🎉 

Minor suggestion:

```suggestion
If you are using DeepSpeed, you can use the following strategies. The usage of the following strategies is the same as listed above, but finetuning with DeepSpeed doesn't yet support the loading and storing of its parameters.

* ``freeze_deepspeed``
* ``no_freeze_deepspeed``
* ``freeze_unfreeze_deepspeed``
* ``unfreeze_milestones_deepspeed``
```

File: flash/core/finetuning.py
Comment (krshrimali) [inline flash/core/finetuning.py:225]:
```suggestion
    DeepSpeed cannot store and load its parameters when working with Lightning. So FlashDeepSpeedFinetuning
    overrides `_store` to not store its parameters.
```

We now have Lightning :)

File: tests/core/test_finetuning.py
Comment (krshrimali) [inline tests/core/test_finetuning.py:266]: Really nice tests! Thanks for adding them.

**Review by krshrimali — APPROVED** [review]
Hi, @ar90n - Thank you so much for working on this PR. I really appreciate that you added the documentation, and relevant tests. 🎉 

LGTM (just added a couple of minor suggestions)!

**Comment by ar90n** [comment]
Hi @krshrimali 
Thanks for your review and suggestions! It's so helpful for me. Because my English is poor. I'm so glad that this PR was approved.

**Comment by ar90n** [comment]
I checked the reason for `Lightning-AI.lightning-flash (Examples)` and found the follwoings.

```
RuntimeError: CUDA error: the launch timed out and was terminated
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

I'm not familiar with them. It seems that the test process was terminated by its timeout.
In my local environment, this test passes. Please some help to solve this issue.

**Comment by krshrimali** [comment]
> I checked the reason for `Lightning-AI.lightning-flash (Examples)` and found the follwoings.
> 
> ```
> RuntimeError: CUDA error: the launch timed out and was terminated
> CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
> For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
> ```
> 
> I'm not familiar with them. It seems that the test process was terminated by its timeout. In my local environment, this test passes. Please some help to solve this issue.

Hi, @ar90n - Please don't worry about it. I'll have to check once on my personal GPU, sometimes these failures can be flaky (because of resources not being available or anything else). I'll merge it once I'm done testing, but this is good to go! 🎉

**Comment by krshrimali** [comment]
The test passes locally on my GPU, let's merge this and monitor the CI. In case an alarm is raised, I'll attempt to fix it. Thanks, @ar90n for your hard-work and patience with this PR. 🎉

**Comment by krshrimali** [comment]
Just added the CHANGELOG entry, let's wait for the CI, and push it ASAP. <3

**Comment by krshrimali** [comment]
@ar90n - FYI, it took us some time to fix the CI, sorry for that. @ethanwharris is currently OOO for this week, so whenever he is back, he'll help merge this. 🎉 Thank you for your contribution, and patience.

**Pull request merged**
